api/src/main/java/io/grpc/InternalConfigSelector.java

```diff
@@ -37,28 +39,51 @@
   public abstract Result selectConfig(LoadBalancer.PickSubchannelArgs args);
 
   public static final class Result {
+    private final Status status;
+    @Nullable
     private final Object config;
+    @Nullable
     private final CallOptions callOptions;
     @Nullable
     private final Runnable committedCallback;
 
-    private Result(Object config, CallOptions callOptions, @Nullable Runnable committedCallback) {
-      this.config = checkNotNull(config, "config");
-      this.callOptions = checkNotNull(callOptions, "callOptions");
+    private Result(
+        Status status, Object config, CallOptions callOptions, Runnable committedCallback) {
+      this.status = checkNotNull(status, "status");
+      this.config = config;
+      this.callOptions = callOptions;
       this.committedCallback = committedCallback;
     }
 
+    /**
+     * Creates a {@code Result} with the given error status.
+     */
+    public static Result forError(Status status) {
+      checkArgument(!status.isOk(), "status is OK");
+      return new Result(status, null, null, null);
+    }
+
+    /**
+     * Returns the status of the config selection operation. If status is not {@link Status#OK},
+     * this result should not be used.
+     */
+    public Status getStatus() {
+      return status;
+    }
+
     /**
      * Returns a parsed config. Must have been returned via
      * ServiceConfigParser.parseServiceConfig().getConfig()
      */
+    @Nullable
     public Object getConfig() {
       return config;
     }
 
     /**
      * Returns a config-selector-modified CallOptions for the RPC.
      */
+    @Nullable
```

Comment:
`@Nullable` is misleading. Although it is nullable from the programing language point of view. The null case is when status is an error you call `getConfig()`, which is wrong usecase. The only true `@Nullable` is `getCommittedCallback()`.

Reply:
From the pure API point of view (just as how a checker framework or IDE uses this annotation), this `Nullable` is legit. For API integrity, nothing can prevent the API consumer calling `getConfig()` even if the Result is with an error status. It would be the consumer's responsibility to use it correctly. But as the API provider, it should tell what is indeed happening.